### PR TITLE
Widen restrictive archive traversal, fixes #1925

### LINKED ISSUE #1925: Nested Archive traversal is too restrictive
The current documentation for the `-nested` flag reads: 

> This option enables or disables scanning of nested jar and zip files found in the list of files and directories to be analyzed. By default, scanning of nested jar/zip files is enabled. To disable it, add `-nested:false` to the command line arguments.

In opposition to that documentation the current implementation in the `ClassPathBuilder` checks examined `ICodeBaseEntry`s against `Archive.isLibraryFileName`:

https://github.com/spotbugs/spotbugs/blob/31375986d170956c5a26f657752879b282601c1f/spotbugs/src/main/java/edu/umd/cs/findbugs/classfile/impl/ClassPathBuilder.java#L701-L703

---

https://github.com/spotbugs/spotbugs/blob/31375986d170956c5a26f657752879b282601c1f/spotbugs/src/main/java/edu/umd/cs/findbugs/util/Archive.java#L74-L77

This clearly contradicts the documentation which states that ".zip" files are also traversed. It also contradicts the underlying assumption of a "nested" flag, which would include traversing **all** java archive types, be that war, ear or sar.

This seems to have been changed back in 2013: c7df6ea

I propose that `isLibraryFileName` be removed, instead deferring to `isArchiveFileName` again (and updating the documentation to reflect that). I unfortunately cannot grasp whether that could have averse effects on the codebase, especially when considering the interplay of classpath and auxclasspath, but currently this is a blocker for me to be able to use spotbugs to analyze my project within a CI pipeline.

### PR DESCRIPTION
**Pull request by Vogel612**
----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Review by ThrawnCA — DISMISSED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/util/Archive.java [hunk 1]
```diff
@@ -70,9 +70,4 @@ private static String getExtension(String fileName) {
         String extension = fileName.substring(lastDot).toLowerCase(Locale.ENGLISH);
         return extension;
     }
-
-    public static boolean isLibraryFileName(String fileName) {
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/util/Archive.java:74]: I think it's not so good to remove a public API by minor/patch update, could you revert this change and add `@Deprecated` annotation instead?

If you make a GitHub issue to remove this deprecated in the next major update, I will add it into the [milestone for v5](https://github.com/spotbugs/spotbugs/milestone/15).

**Review by KengoTODA — CHANGES_REQUESTED** [review]
One request I left, please check. And if possible, try to [reproduce the issue you want to fix by a unit test](https://github.com/spotbugs/spotbugs/blob/31375986d170956c5a26f657752879b282601c1f/.github/CONTRIBUTING.md#before-reporting-a-problem-with-detectors).

**Comment by Vogel612** [comment]
@KengoTODA I have updated the public API as per your request. I am having a hard time finding the existing unit-tests relating to the ClassPathBuilder, though. Could you give me a hint where I should be looking?

**Comment by Vogel612** [comment]
@KengoTODA I added some unit-tests that rely on a handbuilt nested archive I added to the spotbugsTestCases. I really dislike this setup for the test case, but since I am not familiar enough with the codebase (and there was virtually no documentation on how something like this could even be attempted), that's the hack I went with.

Because these tests require acessing the engine configuration and I didn't want to mess with user preferences between test cases I added an overload to the SpotBugsRule used to integrate the Engine into UnitTests to support modifying the engine in code just before the execution of the actual analysis.

**Review by ThrawnCA — APPROVED** [review]

**Review by KengoTODA — CHANGES_REQUESTED** [review]

File: test-harness-core/src/main/java/edu/umd/cs/findbugs/test/AnalysisRunner.java [hunk 2]
```diff
@@ -83,7 +84,7 @@ public AnalysisRunner addAuxClasspathEntry(Path path) {
     }
 
     @Nonnull
-    public BugCollectionBugReporter run(Path... files) {
+    public BugCollectionBugReporter run(Consumer<IFindBugsEngine> engineCustomization, Path... files) {
```
[the remark this answers is not in the document]
Comment (Vogel612) [reply test-harness-core/src/main/java/edu/umd/cs/findbugs/test/AnalysisRunner.java:87]: The only use of that API is from within `SpotBugsRule`. I wouldn't define a testing interface adapter as public API. I had already adapted the API of `SpotBugsRule` to your suggestion, I don't think adding an overload here provides any benefit whatsoever.

**Review by KengoTODA — CHANGES_REQUESTED** [review]
The code is LGTM. Thanks! Add an entry to `CHANGELOG.md` then I'll merge this PR for the coming release.

**Comment by Vogel612** [comment]
@KengoTODA Added a changelog entry and rebased onto the master branch. All commits have been squashed into a single one.

**Review by KengoTODA — APPROVED** [review]
Everything LGTM. Thanks a lot! :+1:

**Review by ThrawnCA — APPROVED** [review]

**Pull request merged**